CMakeLists.txt

```
@@ -0,0 +1,71 @@
+cmake_minimum_required(VERSION 3.13.4)
+
+if(POLICY CMP0116)
+  cmake_policy(SET CMP0116 OLD)
+endif()
+
+project(HEIR LANGUAGES CXX C)
+
+set(CMAKE_CXX_STANDARD 17 CACHE STRING "C++ standard to conform to")
+
+find_package(MLIR REQUIRED CONFIG)
+
+message(STATUS "Using MLIRConfig.cmake in: ${MLIR_DIR}")
+message(STATUS "Using LLVMConfig.cmake in: ${LLVM_DIR}")
+
+option(ENABLE_YOSYS "Enable YOSYS for yosys-optimizer pass (optional)" FALSE)
+
+if(ENABLE_YOSYS)
+   message("Yosys enabled! Make sure ABC and Yosys development files and libraries are on the path")
+else()
+   add_compile_options(-DHEIR_NO_YOSYS=1)
+endif()
+
+
+set(LLVM_RUNTIME_OUTPUT_INTDIR ${CMAKE_BINARY_DIR}/bin)
+set(LLVM_LIBRARY_OUTPUT_INTDIR ${CMAKE_BINARY_DIR}/lib)
+set(MLIR_BINARY_DIR ${CMAKE_BINARY_DIR})
+
+
+
+list(APPEND CMAKE_MODULE_PATH "${MLIR_CMAKE_DIR}")
+list(APPEND CMAKE_MODULE_PATH "${LLVM_CMAKE_DIR}")
+include(TableGen)
+include(AddLLVM)
+include(AddMLIR)
+include(HandleLLVMOptions)
+
+set(INCLUDE_MLIR_LLVM_HEADERS "${CMAKE_BINARY_DIR}/inc/")
+add_custom_command(OUTPUT ${INCLUDE_MLIR_LLVM_HEADERS}
+  COMMAND mkdir -p inc/llvm/include
+  COMMAND mkdir -p inc/mlir/include
+  WORKING_DIRECTORY ${CMAKE_BINARY_DIR}
+  COMMENT "Creating link to LLVM and MLIR headers"
+  VERBATIM
+)
+add_custom_target(HEIRIncludeLLVM
+                  BYPRODUCTS ${CMAKE_BINARY_DIR}/inc/llvm/include/llvm
+                  COMMAND [ -e llvm ] && echo "" || ln -s ${LLVM_INCLUDE_DIRS}/llvm/ ./
+                  DEPENDS ${INCLUDE_MLIR_LLVM_HEADERS}
+                  WORKING_DIRECTORY "${CMAKE_BINARY_DIR}/inc/llvm/include/"
+                  )
+
+add_custom_target(HEIRIncludeMLIR
+                  BYPRODUCTS ${CMAKE_BINARY_DIR}/inc/mlir/include/mlir
+                  COMMAND [ -e mlir ] && echo "" || ln -s ${MLIR_INCLUDE_DIRS}/mlir/ ./
+                  DEPENDS ${INCLUDE_MLIR_LLVM_HEADERS}
+                  WORKING_DIRECTORY "${CMAKE_BINARY_DIR}/inc/mlir/include/")
+message(STATUS "Adding MLIR and LLVM includes.")
+
+include_directories(${CMAKE_BINARY_DIR}/inc)
+include_directories(${LLVM_INCLUDE_DIRS})
+include_directories(${MLIR_INCLUDE_DIRS})
```

This little trick removes the need for the custom command (which unfortunately didn't work for me):
* first set of includes is for heir code (which uses a very long include path for some reason)
* second set is so that includes in included mlir/llvm code (shorter path) work, too
```suggestion
include_directories(${LLVM_INCLUDE_DIRS}/../..)
include_directories(${MLIR_INCLUDE_DIRS}/../..)
include_directories(${LLVM_INCLUDE_DIRS})
include_directories(${MLIR_INCLUDE_DIRS})
```

---

Okay - let me see